source/darwin/dispatch_queue_event_loop.c

```diff
@@ -310,99 +338,110 @@ struct aws_event_loop *aws_event_loop_new_with_dispatch_queue(
+    AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: Starting to destroy Dispatch Queue Event Loop", (void *)event_loop);
+    struct aws_dispatch_loop *dispatch_loop = event_loop->impl_data;
 
-    dispatch_loop->synced_data.is_executing = false;
+    s_lock_synced_data(dispatch_loop);
+    enum aws_dispatch_loop_execution_state execution_state = dispatch_loop->synced_data.execution_state;
+    AWS_FATAL_ASSERT(execution_state == AWS_DLES_RUNNING || execution_state == AWS_DLES_SUSPENDED);
+    if (execution_state == AWS_DLES_SUSPENDED) {
+        dispatch_resume(dispatch_loop->dispatch_queue);
+    }
+    dispatch_loop->synced_data.execution_state = AWS_DLES_SHUTTING_DOWN;
     s_unlock_synced_data(dispatch_loop);
+}
 
-    s_dispatch_event_loop_destroy(dispatch_loop->base_loop);
+static bool s_wait_for_terminated_state(void *user_data) {
+    struct aws_dispatch_loop *dispatch_loop = user_data;
+
+    return dispatch_loop->synced_data.execution_state == AWS_DLES_TERMINATED;
 }
 
-static void s_destroy(struct aws_event_loop *event_loop) {
-    AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: Destroying Dispatch Queue Event Loop", (void *)event_loop);
+static void s_complete_destroy(struct aws_event_loop *event_loop) {
+    AWS_LOGF_TRACE(
+        AWS_LS_IO_EVENT_LOOP, "id=%p: Completing destruction of Dispatch Queue Event Loop", (void *)event_loop);
     struct aws_dispatch_loop *dispatch_loop = event_loop->impl_data;
 
-    /* make sure the loop is running so we can schedule a last task. */
-    s_run(event_loop);
+    // This would be deadlock
+    AWS_FATAL_ASSERT(!aws_event_loop_thread_is_callers_thread(event_loop));
 
     /*
-     * `dispatch_async_and_wait_f()` schedules a block to execute in FIFO order on Apple's dispatch queue and waits
-     * for it to complete before moving on.
-     *
-     * Any block that is currently running or already scheduled on the dispatch queue will be completed before
-     * `s_dispatch_queue_destroy_task()` block is executed.
-     *
-     * `s_dispatch_queue_destroy_task()` will cancel outstanding tasks that have already been scheduled to the task
-     * scheduler and then iterate through cross thread tasks before finally running `s_dispatch_event_loop_destroy()`
-     * which will clean up both aws_event_loop and aws_dispatch_loop from memory.
-     *
-     * It is possible that there are scheduled_iterations that are be queued to run s_run_iteration() up to 1 second
-     * AFTER s_dispatch_queue_destroy_task() has executued. Any iteration blocks scheduled to run in the future will
-     * keep Apple's dispatch queue alive until the blocks complete.
+     * This is the release of the initial ref count of 1 that the event loop was created with.
      */
-    dispatch_async_and_wait_f(dispatch_loop->dispatch_queue, dispatch_loop, s_dispatch_queue_destroy_task);
+    s_dispatch_loop_release(dispatch_loop);
+
+    s_lock_synced_data(dispatch_loop);
+    aws_condition_variable_wait_pred(
```

Comment:
Trivial: May be worth leaving a comment here as there's kind of a lot going on in this one function. 

THIS thread is blocked at this point until the dispatch loop hits zero refs.
synced lock is held but is released by `aws_condition_variable_wait_pred` to allow the lock to be held by the dispatch_loop's event_loop.
synced lock is held again before the `aws_condition_variable_wait_pred` continues at which point we unlock after.

Or not. Mainly had to dig around a bit to make sure we weren't gonna deadlock here as underlying `pthread_cond_wait()` doesn't give much info.

Reply:
Well, we use the condition-variable-and-mutex combo all over the codebase (mostly in tests) so I don't think it's critical to explain a single instance of it.